advtraj/integrate/forward.py

```diff
@@ -1,25 +1,575 @@
+    xarray Dataset
+        Wrapped version of ds_posn.
+
+    """
+    cyclic_coords = ("x", "y")
+    cell_centered_coords = ("x", "y", "z")
+    return wrap_periodic_grid_coords(
+        ds_grid=ds_grid,
+        ds_posn=ds_posn,
+        cyclic_coords=cyclic_coords,
+        cell_centered_coords=cell_centered_coords,
+    )
+
+
+def _calc_backtrack_origin_dist(
+    ds_position_scalars,
+    ds_traj_posn_org,
+    ds_traj_posn,
+    interpolator=None,
+    interp_order=5,
+):
+    """
+    Compute the distance between the true origin and the estimated origin
+    (calculated by back-trajectory from `pt_traj_posn_next`)
+    """
+    ds_grid = ds_position_scalars[["x", "y", "z"]]
+
+    if ds_grid.xy_periodic:
+        ds_traj_posn = _wrap_coords(ds_traj_posn, ds_grid)
+
+    # ds_traj_posn_org contains the 'current' (i.e. known) trajectory
+    # positions.
+
+    # ds_traj_posn contains a guess of the 'next' trajectory position.
+    # Using this estimate of the next trajectory position use the position
+    # scalars at that point to estimate where fluid originated from. If we've
+    # made a good guess for the next position this estimated origin position
+    # should be close to the true origin, i.e. the 'current' trajectory
+    # position.
+    ds_traj_posn_org_guess = calc_trajectory_previous_position(
+        ds_position_scalars=ds_position_scalars,
+        ds_traj_posn=ds_traj_posn,
+        interpolator=interpolator,
+        interp_order=interp_order,
+    )
+
+    ds_p1 = ds_traj_posn_org
+    ds_p2 = ds_traj_posn_org_guess
+    dist_arr = np.array(
+        [
+            ds_p1.x.values - ds_p2.x_est.values,
+            ds_p1.y.values - ds_p2.y_est.values,
+            ds_p1.z.values - ds_p2.z_est.values,
+        ]
+    )
+
+    # Deal with xy wraparound.
+    if ds_grid.xy_periodic:
+        Lx = ds_position_scalars["x"].attrs["Lx"]
+        Ly = ds_position_scalars["y"].attrs["Ly"]
+        xerr = np.asarray(dist_arr[0])
+        xerr[xerr > Lx / 2] -= Lx
+        xerr[xerr < -Lx / 2] += Lx
+        yerr = np.asarray(dist_arr[1])
+        yerr[yerr > Ly / 2] -= Ly
+        yerr[yerr < -Ly / 2] += Ly
+        dist_arr = np.array(
+            [
+                xerr,
+                yerr,
+                dist_arr[2],
+            ]
+        )
+    # Deal with single trajectory case
+    if dist_arr.ndim == 1:
+        dist_arr = np.expand_dims(dist_arr, 1)
+    return dist_arr
+
+
+def _pt_ds_to_arr(ds_pt):
+    return np.array([ds_pt[c].data for c in "xyz"])
+
+
+def _pt_arr_to_ds(arr_pt):
+    # Ensure arr_pt has 2 dimensions.
+    arr_pt = arr_pt.reshape((3, -1))
+    ds_pt = xr.Dataset()
+    for n, c in enumerate("xyz"):
+        ds_pt[c] = xr.DataArray(arr_pt[n, :], dims=("trajectory_number"))
+
+    return ds_pt
+
+
+def _backtrack_origin_point_iterate(
+    ds_position_scalars,
+    ds_traj_posn_org,
+    ds_traj_posn_first_guess,
+    interpolator,
+    solver,
+    interp_order=5,
+    pioptions=None,
```

Comment:
I would make the default value an empty dict here `{}` rather than `None` and turning that into a dict

Reply:
Generally, my understanding is that having mutable defaults in python is dangerous - see e.g.  [https://docs.python-guide.org/writing/gotchas/] and setting to None as default the usin gthat to set an empty dict in th ecode is th erecommendation.